src/ops/fuelup_self.rs

```diff
@@ -1,17 +1,43 @@
+use std::{fs, path::Path};
+use tempfile::tempdir;
+use tracing::error;
+
 use anyhow::Result;
 
 use crate::{
     component,
     download::{download_file_and_unpack, unpack_bins, DownloadCfg},
-    path::fuelup_bin_dir,
+    path::{fuelup_bin, fuelup_bin_dir},
 };
 
+pub fn attempt_install_self(download_cfg: DownloadCfg, dst: &Path) -> Result<()> {
+    download_file_and_unpack(&download_cfg, dst)?;
+    unpack_bins(dst, dst)?;
+
+    Ok(())
+}
+
 pub fn self_update() -> Result<()> {
     let download_cfg = DownloadCfg::new(component::FUELUP, None)?;
-    let fuelup_bin_dir = fuelup_bin_dir();
+    let fuelup_bin = fuelup_bin();
+    let temp_dir = tempdir()?;
+    let fuelup_backup = temp_dir.path().join("fuelup-backup");
+
+    if fuelup_bin.exists() {
+        // Make a backup of fuelup.
+        fs::copy(&fuelup_bin, &fuelup_backup).expect("Could not make a fuelup backup");
+        // We cannot copy new fuelup over the old; we must unlink it first.
+        fs::remove_file(&fuelup_bin).expect("Could not remove fuelup");
+    };
+
+    if let Err(e) = attempt_install_self(download_cfg, &fuelup_bin_dir()) {
+        error!("Failed to install and replace fuelup. {}", e);
 
-    download_file_and_unpack(&download_cfg, &fuelup_bin_dir)?;
-    unpack_bins(&fuelup_bin_dir, &fuelup_bin_dir)?;
+        // We want to restore the backup fuelup in case download goes wrong.
+        if fuelup_backup.exists() {
+            fs::copy(&fuelup_backup, &fuelup_bin).expect("Could not restore fuelup-backup");
```

Comment:
While this error path should be rare, we should return an `Err` that also recommends reinstalling `fuelup` as if this fails we're left with no `fuelup` bin on `PATH`.

Reply:
Upon success, we might want to cleanup the `fuelup_backup` too just because we can:

```rust
// Ignore the error case - this was located in a `temp` directory anyway.
let _ = fs::remove_file(&fuelup_backup);
```